File: clojure/matching-brackets/_iterations/1/src/matching_brackets.clj
(ns matching-brackets
  (:require [clojure.string :as string]))

(def ^:private closing
  { \{ \}
    \[ \]
    \( \) })

(defn valid? [expression]
  (let [sanitized (string/replace expression #"[^\[\]{}()]+" "")]
    (loop [[c & remain] sanitized
           stack '()]
      (if-not c
        (empty? stack)
        (let [opposite (get closing c)
              closing? (nil? opposite)
              matches? (= (first stack) c)]
          (cond
            (and closing? (not matches?)) false
            (and closing? matches?)       (recur remain (rest stack))
            :else                         (recur remain (conj stack opposite))))))))

File: ruby/matching-brackets/_iterations/1/matching_brackets.rb
module Brackets
  def self.paired?(input)
    brackets = input.gsub(/[^\[\]\{\}\(\)]/, "")
    while !brackets.empty?
      new_brackets = brackets.gsub(/\[\]|\{\}|\(\)/, "")
      return false if new_brackets == brackets
      brackets = new_brackets
    end
    true
  end
end
